man/copris.1
.\" Automatically generated by Pandoc 3.8.3
.\"
.TH "COPRIS" "1" "2026\-04\-28" "" "0.9\-781\-g611e6b2\-dirty"
.SH NAME
\f[B]copris\f[R] \- a converting printer server
.SH SYNOPSIS
.PP
\f[B]copris\f[R] [\f[I]options\f[R]] [\f[I]printer device\f[R] or
\f[I]output file\f[R]]
.PD 0
.P
.PD
\f[B]copris\f[R] \f[B]\-\-dump\-commands\f[R] > new_feature_file.ini
.SH DESCRIPTION
COPRIS is a printer server and text conversion program that bridges the
gap between modern, UTF\-8\-encoded, Markdown\-formatted text and
(older) dot\-matrix and thermal printers.
.PP
When provided with the appropriate configuration files and options,
COPRIS reads plain text from a local source or the network (with a
built\-in TCP socket server), applies optional processing methods to it
and sends it to the destination device (or file).
.PP
Two main text processing methods stand out in COPRIS; they are:
.SS Character recoding via encoding files
Older printers usually don\(cqt support multi\-byte UTF\-8\-encoded
text.
Rather, they use code pages and customised character sets which may vary
between different manufacturers.
COPRIS can load a character \f[I]encoding file\f[R] that is used to
recode received text.
By writing or selecting a premade one for your printer and locale, you
allow COPRIS to recode (replace) received characters with appropriate
locale\-specific ones, understood by your printer.
.PP
This is particularly useful in cases where printer distributors swapped
lesser\-used ASCII characters (such as \f[CR]\(ti \(at {\f[R]) with
those from the national alphabet, so they could be typed out directly
from the non\-localised keyboard.
By finding similar\-looking characters in printer\(cqs character set,
you may, crudely, regain those lost characters.
.PP
However, \f[I]encoding files\f[R] aren\(cqt tied only to locale
conversions, you may use them to remap any character.
Examples include typographic quotation marks, em dashes, copyright
symbol and others, commonly found in character sets of various printers.
.SS Markdown formatting via printer feature files
Many printers support commands (escape codes) for formatting printed
text (e.g.\ Epson has its ESC/P control language).
COPRIS can read a \f[I]printer feature file\f[R] that specifies such
commands, invoked by following Markdown attributes:
.IP \(bu 2
bold and italic \f[I]emphasis\f[R] via asterisks (\f[CR]*\f[R],
\f[CR]**\f[R], \f[CR]***\f[R]) or underscores (\f[CR]_\f[R],
\f[CR]__\f[R], \f[CR]___\f[R])
.IP \(bu 2
up to 4 levels of \f[I]headings\f[R] via number signs (\f[CR]#\f[R],
\f[CR]##\f[R], \f[CR]###\f[R], \f[CR]####\f[R])
.IP \(bu 2
\f[I]blockquotes\f[R] via greater\-than signs and a space
(\f[CR]>\ \f[R])
.IP \(bu 2
\f[I]inline code\f[R] via single backticks (\f[CR]\(ga\f[R])
.IP \(bu 2
\f[I]code blocks\f[R] via triple backticks (\f[CR]\(ga\(ga\(ga\f[R])
\(en or four spaces, but those are disabled by default
.IP \(bu 2
\f[I]links\f[R] via angle brackets (\f[CR]< >\f[R])
.PP
Any command can be bound to mentioned attributes.
If desired, bold emphasis can trigger double\-width text, headings may
change the font face or size and links can be made underlined.
There are no limits, as long as your \f[I]printer feature file\f[R]
contains commands that are understood by the printer (or any other
destination device).
.PP
You might have noticed that only a subset of the Markdown specification
is supported.
COPRIS intentionally doesn\(cqt alter text layout (white space,
paragraphs etc.), since there\(cqs no markup/layout engine.
You may freely format your input plain\-text files to your liking, and
COPRIS will preserve the spacing and newlines (just be warned about four
leading spaces triggering a code block).
.SH FILE FORMAT
\f[I]Encoding\f[R] and \f[I]printer feature files\f[R] are expected to
be in the INI file format:
.IP
.EX
\f[I]# Comment\f[R]
key1 = whitespace separated values
key2 = whitespace separated values ; inline comment
\&...
.EE
.PP
Left\-hand keys should be written out directly as UTF\-8 characters or
strings, right\-hand values are expected to be in either decimal, octal
(prefixed by a \f[CR]0\f[R]) or hexadecimal base (prefixed with
\f[CR]0x\f[R]).
.PP
No indentation is permitted before writing out left\-hand keys.
White space at beginning of lines will result in (sometimes) confusing
error messages \- unfortunately a limitation of the parsing library we
use.
.PP
However, any amount of spaces (or tabulators) are allowed after that.
Keep in mind not to prepend zeroes to decimal numbers for alignment, as
they\(cqll be interpreted as octal numbers.
.PP
One line can contain exactly \f[I]one\f[R] definition/command.
.PP
In case you need a \f[CR]key\f[R] to resolve to a blank
\f[CR]value\f[R], set \f[CR]\(at\f[R] as the value (e.g.\ if you want to
remove certain characters from the text, or omit rendering some Markdown
attributes).
.PP
Here are some examples:
.IP
.EX
Å = 0xC5                ; for an encoding file
F_BOLD_OFF = 0x1B 0x45  ; for a printer feature file
F_H3_ON = \(at             ; ditto
  NOT_ALLOWED = 7       ; forbidden leading spaces
.EE
.SH THE ENCODING FILE
Before making one, check if COPRIS already includes one for your locale
(\f[CR]encodings\f[R] directory in the source tree).
.PP
The format of an encoding file is as follows.
Start the line by entering your desired character, either via the
keyboard or copied from somewhere else.
Follow by an equals sign.
Then, enter the numerical value of the same character your printer will
understand.
.PP
Here\(cqs an example excerpt of the YUSCII encoding:
.IP
.EX
\f[I]# Left\-hand letters are readable to us, however, their UTF\-8 codes\f[R]
\f[I]# are unknown to the printer. Therefore, COPRIS replaces them with\f[R]
\f[I]# the right\-hand values, here represented in hexadecimal.\f[R]
Č = 0x5E  ; replacing \(ha
Ž = 0x40  ; replacing \(at
Š = 0x7B  ; replacing [
.EE
.PP
You might have noticed that YUSCII overrides some ASCII characters.
Luckily, printer character sets often consist of additional characters
(Greek letters, box drawing and maths symbols etc.), which can be also
specified in an encoding file to imitate the lost ones:
.IP
.EX
\(ha = 0x27  ; replacement is an acute accent
\(at = 0xE8  ; replacement is the Greek letter Phi
\&...
.EE
.PP
Have you transcribed a national or a well known encoding COPRIS
doesn\(cqt yet include?
Contributions are always welcome!
.PP
You may find out that your printer supports a code page, appropriate for
your locale, but it isn\(cqt selected by default when you turn it on.
Read on, the chapter about printer feature files explains session
commands which can be used for setting up the printer.
.SH THE PRINTER FEATURE FILE
As with encoding files, COPRIS\(cq source tree might have the right one
in the \f[CR]feature\-files/\f[R] directory.
If the file name seems appropriate, check the first few lines of the
file to see if your printer/use case is supported.
.PP
Else, consult the printer\(cqs manual.
There should be a section on escape codes.
Generate a sample printer feature file
(\f[CR]copris \-\-dump\-commands > my\-printer.ini\f[R]).
Uncomment the appropriate commands and append values from the manual, in
the same way as with encoding files.
.PP
Example from Epson\(cqs LX\-300 manual, page A\-14 (91):
.IP
.EX
ASCII   Dec.   Hex.   Description
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
ESC 4   52     34     Select Italic Mode
ESC 5   53     35     Cancel Italic Mode
.EE
.PP
The corresponding printer feature file lines are:
.IP
.EX
\f[I]# lx300.ini\f[R]
F_ITALIC_ON  = 0x1B 0x34  ; hexadecimal notation, 0x1B = ESC
F_ITALIC_OFF = 27 53      ; decimal notation, 27 = ESC
.EE
.SS Variables in printer feature files
You can use existing command names as variables, as long as you define
the command \f[I]before\f[R] using it as a variable.
Furthermore, you may define your own custom variables and use them in
existing commands.
For COPRIS to recognise them, they must be prefixed with \f[CR]C_\f[R]!
Variables may be interweaved with commands.
.PP
Examples:
.IP
.EX
\f[I]# lx300.ini \- continued\f[R]
C_UNDERLINE_ON  = 0x1B 0x2D 0x31
C_UNDERLINE_OFF = 0x1B 0x2D 0x30
F_H1_ON  = C_UNDERLINE_ON F_ITALIC_ON
F_H1_OFF = F_ITALIC_OFF C_UNDERLINE_OFF
.EE
.SS Session commands
COPRIS provides two command pairs for sending repetitive settings to the
printer.
They may be used to set the code page, text margins, line spacing, font
face, character density, initialise/reset the printer and so on:
.IP \(bu 2
\f[CR]S_AT_STARTUP\f[R] and \f[CR]S_AT_SHUTDOWN\f[R] \- sent to the
printer once after COPRIS starts and once before it exits
.IP \(bu 2
\f[CR]S_BEFORE_TEXT\f[R] and \f[CR]S_AFTER_TEXT\f[R] \- sent to the
printer each time text is received, in order \f[CR]S_BEFORE_TEXT\f[R] \-
\f[I]received text\f[R] \- \f[CR]S_AFTER_TEXT\f[R]
.PP
Examples:
.IP
.EX
\f[I]# lx300.ini \- continued\f[R]
S_AT_STARTUP = 0x07  ; sound the bell (beeper)
S_AFTER_TEXT = 0x0C  ; trigger form feed after printing
.EE
.SH VARIABLES, NUMERICAL VALUES AND COMMENTS
If you\(cqve defined a printer feature file, and it includes custom
variables (prefixed with \f[CR]C_\f[R]), you may use them directly in
input text.
.PP
To allow for that, begin your text with a line, containing
\f[CR]COPRIS ENABLE\-VARIABLES\f[R].
This is called the \f[I]modeline\f[R] and tells COPRIS to process
variables in input text.
.PP
You may then call variables in the text file by prepending a dollar sign
to them.
I.e., if your variable is \f[CR]C_SERIF\f[R], \f[CR]$C_SERIF\f[R] should
be used in text.
If you want to call multiple, write them out space\-separated.
Note that anything between the dollar sign and the new line is
interpreted as a variable (or a series of them).
.PP
Furthermore, apart from already\-defined variables, numerical values can
be included in text.
They must be written out in the same way as variables \- prefixed with a
dollar sign, and, if specifying a multitude of them, separated by white
spaces.
They must be in either decimal, octal or hexadecimal notation, as they
would be in a printer feature file.
.PP
Lastly, comments can be passed in text.
They again consist of a dollar sign, which is then followed by a number
sign and your comment message.
.PP
Here are examples of all three notations:
.IP
.EX
COPRIS ENABLE\-VARIABLES    (must be included at the top)
$# This reduces line spacing:
$C_ESC 0x33 25             (feature file has a C_ESC command defined)
.EE
.SS Mixing variables and text
Be wary of mixing variables and text in a sentence.
Variables will only get detected properly if they\(cqre at the end of
the sentence, not followed by any punctuation.
.PP
Meaning:
.IP
.EX
Word count is $C_WC        (works normally)
The sum is $C_WC words.    (fails \- expects a variable named \(aq$C_WC\ words.\(aq)
.EE
.PP
For reliable interpretation of variables, dedicate a whole line just for
them, and make use of Markdown for mixing text and commands.
.PP
COPRIS will skip parsing variables without the correct prefix (letter
and underscore) and copy them out to the output.
If you\(cqd like to do the same to a variable with one of the correct
prefixes, escape it by prepending it another dollar symbol.
.SS The modeline
It is expected in the first line of received text.
Its format is:
.IP
.EX
COPRIS <required 1st option> [ optional 2nd option ]
.EE
.PP
Previous documentation already mentioned its use for for enabling
variables.
It serves one other purpose, disabling parsing Markdown in text.
.PP
Here are both modeline options and their short forms:
.IP \(bu 2
\f[CR]ENABLE\-VARIABLES\f[R]; \f[CR]ENABLE\-VARS\f[R]
.IP \(bu 2
\f[CR]DISABLE\-MARKDOWN\f[R]; \f[CR]DISABLE\-MD\f[R]
.PP
Letters are case\-insensitive and the order of options isn\(cqt
important.
Thus, following lines are the same:
.IP \(bu 2
\f[CR]COPRIS ENABLE\-VARIABLES DISABLE\-MARKDOWN\f[R]
.IP \(bu 2
\f[CR]copris disable\-md enable\-vars\f[R]
.SH COMMAND LINE OPTIONS
.TP
\f[B]\-p\f[R], \f[B]\-\-port\f[R] \f[I]NUMBER\f[R]
Run COPRIS as a network server on port \f[I]NUMBER\f[R].
Superuser privileges are required if \f[I]NUMBER\f[R] is less than 1024.
.TP
\f[B]\-e\f[R], \f[B]\-\-encoding\f[R] \f[I]FILE\f[R]
Recode characters in received text according to definitions from
encoding \f[I]FILE\f[R].
This option can be specified multiple times with different
\f[I]FILE\f[R]s.
Consult the \f[B]FILE FORMAT\f[R] and \f[B]THE ENCODING FILE\f[R]
chapters above for details.
.TP
\f[B]\-\-ignore\-missing\f[R]
If recoding characters, do not terminate the program if received text
contains any possibly unwanted multi\-byte characters that were not
handled by specified \f[I]FILE\f[R]s.
.TP
\f[B]\-f\f[R], \f[B]\-\-feature\f[R] \f[I]FILE\f[R]
Process Markdown markup and variables in received text and apply session
commands according to printer feature \f[I]FILE\f[R].
This option can be specified multiple times with different
\f[I]FILE\f[R]s.
Consult the \f[B]FILE FORMAT\f[R], \f[B]THE PRINTER FEATURE FILE\f[R]
and \f[B]VARIABLES, NUMERICAL VALUES AND COMMENTS\f[R] chapters above
for details.
.TP
\f[B]\-\-dump\-commands\f[R]
Show all possible printer feature commands in INI file format (e.g.\ to
be piped into a new printer feature file you are making).
.TP
\f[B]\-d\f[R], \f[B]\-\-daemon\f[R]
If running as a network server, do not exit after the first connection.
.TP
\f[B]\-l\f[R], \f[B]\-\-limit\f[R] \f[I]NUMBER\f[R]
If running as a network server, limit number of received bytes to
\f[I]NUMBER\f[R].
.TP
\f[B]\-\-cutoff\-limit\f[R]
If limit is active, cut text on \f[I]NUMBER\f[R] count instead of
discarding the whole chunk.
.TP
\f[B]\-v\f[R], \f[B]\-\-verbose\f[R]
Show informative status messages.
If specified twice, show even more messages.
.TP
\f[B]\-q\f[R], \f[B]\-\-quiet\f[R]
Do not show any unnecessary messages, except warnings and fatal errors,
routed to \f[I]stderr\f[R].
This also omits \f[I]notes\f[R], shown if COPRIS assumes it is not
invoked properly.
.TP
\f[B]\-h\f[R], \f[B]\-\-help\f[R]
Show a short option summary.
.TP
\f[B]\-V\f[R], \f[B]\-\-version\f[R]
Show program version, author and build\-time options.
.PP
Do not specify a port number if you want to read from standard input.
Likewise, omit the output file to have text echoed out to standard
output (or piped elsewhere).
.SH EXAMPLES OF INVOKING COPRIS
\f[B]Notice:\f[R] COPRIS is in active development.
Some features are still missing, others have not been thoroughly tested
yet.
Command line option arguments may change in future.
Version 1.0 will be tagged when the feature set will be deemed complete.
.PP
Run as a simple server on port 8080, perform no text recoding, output
received data to the serial port and exit after one connection.
Note that superuser privileges are required if the specified port is
smaller than 1024.
.IP
.EX
copris \-p 8080 /dev/ttyS0
.EE
.PP
Serve on port 8080 as a daemon (do not exit after first connection),
recode text using the \f[CR]slovene.ini\f[R] encoding file, limit any
incoming text to a maximum of 100 characters and print received data to
the terminal.
Note that text limit works only when running as a server.
.IP
.EX
copris \-p 8080 \-d \-e slovene.ini \-l 100
.EE
.PP
Read local file \f[CR]font\-showcase.md\f[R] using the printer feature
file \f[CR]epson\-escp.ini\f[R].
Interpret any possible user commands, found in the local file.
Output formatted text to an USB printer interface on the local computer:
.IP
.EX
copris \-f epson\-escp.ini \-c /dev/usb/lp0 < font\-showcase.md
.EE
.SH KNOWN DEFICIENCIES
.IP \(bu 2
no warning is produced if there\(cqs a typo in the modeline, or an
unknown option mixed with valid ones
.IP \(bu 2
DOS/Windows style line endings (\f[CR]\(rsr\(rsn\f[R]) aren\(cqt
supported \- process input text through \f[I]dos2unix(1)\f[R] if
necessary
.IP \(bu 2
markup in Markdown quotes is ignored
.IP \(bu 2
since signals aren\(cqt caught, COPRIS doesn\(cqt exit cleanly when
interrupted
.IP \(bu 2
when sending data over the network, not all errors are sent back to the
sender
.SH DEVELOPMENT
COPRIS\(cq development repository resides at \c
.UR https://github.com/bertronika/copris
.UE \c
\&.
.SH SEE ALSO
\f[B]stty\f[R](1), \f[B]intercopris\f[R](1)


src/copris.c
/*
 * copris.c
 * Parses arguments, trfile and printerset selections and runs the server
 * 
 * COPRIS - a converting printer server
 * (c) 2020-21 Nejc Bertoncelj <nejc at bertoncelj.eu.org>
 * 
 * Licensed under the MIT/Expat license.
 * 
 * - https://azrael.digipen.edu/~mmead/www/Courses/CS180/getopt.html
 */

#define COPRIS_VER "0.9"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <limits.h>

#include "common.h"
#include "config.h"
#include "debug.h"
#include "copris.h"
#include "server.h"
#include "translate.h"
#include "printerset.h"

#define list_prsets() for(int p = 0; printerset[p][0][0] != '\0'; p++)\
	printf("%s  ", printerset[p][0]);

/* 
 * Verbosity levels:
 * 0  silent/fatal
 * 1  error
 * 2  info
 * 3  debug
 */
int verbosity = 1;

int main(int argc, char **argv) {
	int opt;               // Character, read by getopt
	char *parserr;         // String to integer conversion error
	int terminate = 0;     // 1 -> free pointers, terminate
	                       // 2 -> free pointers, display help (3 -> version)
	int daemon    = 0;     // Is the daemon option set?
	int is_stdin  = 0;     // Don't open a socket if true

	int parentfd  = 0;     // Parent file descriptor to hold a socket
	int portno    = -1;    // Listening port of this server

	int limitnum     = 0;  // Limit received number of bytes
	int limit_cutoff = 0;  // Cut text off at limit instead of discarding it

	struct attrib trfile, prset, destination;
	trfile      = (struct attrib) { 0 };  // Translation file (user-supplied)
	prset       = trfile;                 // Printer set (compiled in)
	destination = trfile;                 // Destination filename

	/* man 3 getopt_long */
	static struct option long_options[] = {
		{"port",          1, NULL, 'p'},
		{"daemon",        0, NULL, 'd'},
		{"trfile",        1, NULL, 't'},
		{"printer",       1, NULL, 'r'},
		{"limit",         1, NULL, 'l'},
		{"cutoff-limit",  0, NULL, 'D'},
		{"verbose",       2, NULL, 'v'},
		{"quiet",         0, NULL, 'q'},
		{"help",          0, NULL, 'h'},
		{"version",       0, NULL, 'V'},
		{NULL,            0, NULL,   0} /* end-of-list */
	};
	
	// int argc, char* argv[], char* optstring, struct* long_options, int* longindex
	while((opt = getopt_long(argc, argv, ":p:dt:r:l:vqhV", long_options, NULL)) != -1 &&
		  !terminate) {
		switch(opt) {
		case 'p':
			portno = strtol(optarg, &parserr, 10);
			if(*parserr) {
				fprintf(stderr, "Unrecognised characters in port number (%s). "
				                "Exiting...\n", parserr);
				terminate = 1;
				break;
			}

			if(portno > 65535 || portno < 1) {
				fprintf(stderr, "Port number out of range. "
				                "Exiting...\n");
				terminate = 1;
				break;
			}
			break;
		case 'd':
			daemon = 1;
			break;
		case 't':
			if(strlen(optarg) <= FNAME_LEN) {
				terminate = store_argument(optarg, &trfile);
			} else {
				fprintf(stderr, "Trfile filename too long (%s). "
				                "Exiting...\n", optarg);
				terminate = 1;
			}
			break;
		case 'r':
			if(strlen(optarg) <= PRSET_LEN) {
				terminate = store_argument(optarg, &prset);
			} else {
				// Excessive length already makes it wrong
				fprintf(stderr, "Selected printer feature set does not exist (%s). "
				                "Exiting...\n", optarg);
				terminate = 1;
			}
			break;
		case 'l':
			limitnum = strtol(optarg, &parserr, 10);
			if(*parserr) {
				fprintf(stderr, "Unrecognised characters in limit number (%s). "
				                "Exiting...\n", parserr);
				terminate = 1;
				break;
			}

			if(limitnum > 4096 || limitnum < 0) {
				fprintf(stderr, "Limit number out of range. "
				                "Exiting...\n");
				terminate = 1;
				break;
			}
			break;
		case 'D':
			limit_cutoff = 1;
			break;
		case 'v':
			if(verbosity < 3)
				verbosity++;
			break;
		case 'q':
			verbosity = 0;
			break;
		case 'h':
			terminate = 2;
			break;
		case 'V':
			terminate = 3;
			break;
		case ':':
			if(optopt == 'p')
				fprintf(stderr, "Port number is missing. "
				                "Exiting...\n");
			else if(optopt == 't')
				fprintf(stderr, "Translation file is missing. "
				                "Exiting...\n");
			else if(optopt == 'r')
				fprintf(stderr, "Printer set is missing. "
				                "Exiting...\n");
			else if(optopt == 'l')
				fprintf(stderr, "Limit number is missing. "
				                "Exiting...\n");
			else
				fprintf(stderr, "Option '-%c' is missing an argument. "
				                "Exiting...\n", optopt);
			terminate = 1;
			break;
		case '?':
			if(optopt == 0)
				fprintf(stderr, "Option '%s' not recognised. "
				                "Exiting...\n", argv[optind - 1]);
			else
				fprintf(stderr, "Option '-%c' not recognised. "
				                "Exiting...\n" , optopt);
			terminate = 1;
			break;
		default:
			fprintf(stderr, "Undefined problem while parsing options. "
			                "Exiting... \n");
			terminate = 10;
			break;
		}
	}
	
	// Only one destination argument is accepted, others are discarded
	if(!terminate && argv[optind]) {
		if(strlen(argv[optind]) <= FNAME_LEN) {
			terminate = store_argument(argv[optind], &destination);
		} else {
			fprintf(stderr, "Destination filename too long (%s). " 
			                "Exiting...\n", argv[optind]);
			terminate = 1;
		}

		if(access(destination.text, W_OK) == -1) {
			log_perr(-1, "access", "Unable to write to output file/printer. Does "
			                       "it exist, with appropriate permissions?");

			terminate = 1;
		}
	}
	
	if(!terminate && prset.exists) {
		prset.exists = 0;
		for(int p = 0; printerset[p][0][0] != '\0'; p++) {
			if(strcmp(prset.text, printerset[p][0]) == 0) {
				prset.exists = p + 1;
				break;
			}
		}

		if(prset.exists < 1) {
			fprintf(stderr, "Selected printer feature set does not exist "
			                "(%s). ", prset.text);
			if(verbosity) {
				fprintf(stderr, "Exiting...\n");
				terminate = 1;
			} else {
				fprintf(stderr, "Disabling it.\n");
			}
		}

		free(prset.text);
	}

	// Check and load translation definitions
	if(!terminate && trfile.exists && copris_trfile(trfile.text)) {
		trfile.exists = 0; // trfile is free'd in copris_trfile
		if(verbosity) {
			// Error in trfile. We are not quiet, so notify and exit
			fprintf(stderr, "Exiting...\n");
			terminate = 1;
		} else {
			// Error in trfile. We are quiet, so disable, notify and continue
			fprintf(stderr, "Disabling translation.\n");
		}
	}

	// If an erroneous argument is passed after allocations, free them
	if(terminate) {
		if(trfile.exists)
			free(trfile.text);

		if(prset.exists)
			free(prset.text);

		if(destination.exists)
			free(destination.text);

		if(terminate == 2) {
			copris_help(argv[0]);
			return 0;
		} else if(terminate == 3) {
			copris_version();
			return 0;
		}

		return terminate;
	}

	if(argc < 2 && log_err()) {
		if(log_info())
			log_date();
		else
			printf("Note: ");

		printf("Without any arguments, COPRIS won't do much. "
               "Try using the '--help' option.\n");
	}

	if(portno < 1)
		is_stdin = 1;

	if(log_info()) {
		log_date();
		printf("Verbosity level set to %d.\n", verbosity);
	}

	if(daemon && is_stdin) {
		daemon = 0;
		if(log_err()){
			if(log_info())
				log_date();
			else
				printf("Note: ");
			
			printf("Daemon mode not available while reading from stdin.\n");
		}
	}

	if(daemon && log_debug()) {
		log_date();
		printf("Daemon mode enabled.\n");
	}
	
	if(prset.exists && log_info()) {
		log_date();
		printf("Selected printer feature set %s.\n",
		       printerset[prset.exists - 1][0]);
	}
	
	if(limitnum > 0 && log_debug()) {
		log_date();
		printf("Limiting received number of bytes to %d B.\n", limitnum);
	}
	
	if(log_debug() && !is_stdin) {
		log_date();
		printf("Server listening port set to %d.\n", portno);
	}
	
	if(log_info()) {
		log_date();
		printf("Data stream will be sent to ");
		if(destination.exists)
			printf("%s.\n", destination.text);
		else
			printf("stdout.\n");
	}
	
	// Open socket and listen
	if(!is_stdin)
		if((terminate = copris_listen(&parentfd, portno)))
			return terminate;

	do {
		if(!is_stdin) {
			// Read from socket
			terminate = copris_read(&parentfd, daemon, limitnum, limit_cutoff,
			                        &trfile, prset.exists, &destination);
		} else {
			// Read from stdin
			terminate = copris_stdin(&trfile, prset.exists, &destination);
		}
	} while(daemon && !terminate);

	if(destination.exists)
		free(destination.text);

	if(terminate)
		return terminate;

	if(log_debug() && !is_stdin) {
		log_date();
		printf("Daemon mode is not set, exiting.\n");
	}

	return 0;
}

int store_argument(char *optarg, struct attrib *attribute) {
	attribute->text = malloc(strlen(optarg) + 1);

	if(!attribute->text) {
		log_perr(-1, "malloc", "Memory allocation error.");
		return 1;
	} else {
		attribute->exists = 1;
		strcpy(attribute->text, optarg);
	}

	return 0;
}

void copris_help(char *copris_location) {
	printf("Usage: %s [arguments] <printer location>\n\n"
	       "  -p, --port NUMBER      Listening port\n"
	       "  -t, --trfile TRFILE    Character translation file\n"
	       "  -r, --printer PRSET    Printer feature set\n"
	       "  -d, --daemon           Run as a daemon\n"
	       "  -l, --limit NUMBER     Limit number of received bytes\n"
	       "      --cutoff-limit     Cut text off at limit instead of\n"
	       "                         discarding the whole chunk\n"
	       "\n"
	       "  -v, --verbose          Be verbose (-vv more)\n"
	       "  -q, --quiet            Display nothing except fatal errors (to stderr)\n"
	       "  -h, --help             Show this help\n"
	       "  -V, --version          Show program version and included printer\n"
	       "                         feature sets\n"
	       "\n"
	       "Printer location can either be an actual printer address, such\n"
	       "as /dev/ttyS0, or a file. If left empty, output is printed to stdout.\n"
	       "\n"
	       "Data can also be read from stdin. If this is desired, omit the port\n"
	       "argument, as incoming connections are prefered over stdin.\n"
	       "\n"
	       "Notes are only printed at situations where COPRIS assumes it is not\n"
	       "invoked correctly. Be sure to omit the quiet argument when debugging.\n",
	       copris_location);
}

void copris_version() {
	printf("COPRIS version %s%s\n"
	       "(C) 2020-21 Nejc Bertoncelj <nejc at bertoncelj.eu.org>\n\n"
	       "Compiled options:\n"
	       "  Buffer size:          %4d bytes\n"
	       "  Max. filename length: %4d characters\n",
	       COPRIS_VER, COPRIS_RELEASE, BUFSIZE, FNAME_LEN);
	printf("Included printer feature sets:\n  ");
	list_prsets();
	printf("\n");
}
